crates/turborepo-lib/src/task_graph/visitor/exec.rs

```diff
@@ -455,7 +463,17 @@ impl ExecContext {
             // Something else killed the child
             ChildExit::KilledExternal => Err(InternalError::ExternalKill),
             // The child was killed by turbo indicating a shutdown
-            ChildExit::Killed | ChildExit::Interrupted => Ok(ExecOutcome::Shutdown),
+            ChildExit::Killed | ChildExit::Interrupted => {
+                // We distinguish between a full shutdown and a restart based on whether the
+                // process manager is closing. If it is closing, it means we are shutting down
+                // the entire run. If it is not closing, it means we are restarting specific
+                // tasks.
+                if self.manager.is_closing() {
```

Comment:
This is potentially dangerous based on race conditions and dependent task behavior. What tests can we write to validate that we're not creating bugs here?

Reply:
If we introduce some atomics, this concern would get addressed.

Reply:
I added an `is_restarting` atomic to the child and check that as the condition instead. The code guarantees that:

- if `stop_tasks` (for a restart) is called first, the child will always return `is_restarting = true` (so this child will consistently return `ExecOutcome::Restarted`)
- if `stop` (for shutdown) is called first, subsequent calls to `stop_tasks` will not set the `is_restarting` flag (so this child will consistently return `ExecOutcome::Shutdown`)